[nnc] Add simplifierUnderContext for simplification that needs context info: currently added for-stmt index var bounds info as context

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62138
* #60693
* #60688
* __->__ #60687
* #60686

Differential Revision: [D29373315](https://our.internmc.facebook.com/intern/diff/D29373315)

huiguoo committed:
[nnc] Add simplifierUnderContext for simplification that needs context info: currently added for-stmt index var bounds info as context

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373315).

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2404,6 +2404,92 @@ Stmt* TermExpander::mutate(const Block* v) {
+  const Var* var_key = dynamic_cast<const Var*>(var);
+  auto got = var_bound_info_.find(var_key);
+  if (got != var_bound_info_.end()) {
+    has_bounds = true;
```
Isn't every loop variable unique, even though their "name" could match? When is this case possible?

huiguoo replied:
I was considering more context-sensitive simplifications in nested `cond` and `for` stmts. For example, 
```
if (i>=0 && i<5) {
  for (i=0; i<3; i++){
    A[i] = ...
  }
  x = (i+20) / 5;
}
```
Inside the `if` stmt, `i` is in the range of [0, 5); and if we can restore this bound info after the `for` stmt, we can use it to simplify the assignment stmt `x = (i+20)/5` to `x = 4`.

This (simplifications inside `if` stmt) is not supported yet but I expect to gradually support such cases.

navahgar replied:
Can you please add this explanation as a comment here? Otherwise, it is difficult to understand why we are doing this.

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2404,6 +2404,92 @@ Stmt* TermExpander::mutate(const Block* v) {
+      std::make_pair(start_new, stop_new);
+  var_bound_info_[var_key] = bound_new;
+
+  const Expr* loops = new Sub(stop_new, start_new);
```
nit: s/loops/iters/

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2404,6 +2404,92 @@ Stmt* TermExpander::mutate(const Block* v) {
+        var_bound_info_.erase(var_key);
+      }
+
+      return body_new;
```
Shouldn't `body_new` be mutated by `this` before returning here?

huiguoo replied:
It has been mutated on line `2442`?

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2404,6 +2404,92 @@ Stmt* TermExpander::mutate(const Block* v) {
 
+// SimplifierUnderContext
+
+Stmt* SimplifierUnderContext::mutate(const For* v) {
```
Can you explain the overall idea behind this function and what it tries to accomplish?

huiguoo replied:
This function tries to record the bounds info of index vars in `for` stmts and those info will be used later when simplifying expressions with the index vars. An example is, 
```
for (i=0; i<5; i++){
  x = (i + 20) % 5;
}
``` 
Without any information about `i`, we are not able to simply the assignment stmt any further; but with the bounds information provided in the `for` stmt, `(i + 20) % 5` can be simplified to `i`. I have seen such stmts in real case. 

Back to your question, the `mutate` function here records the bounds info of `i`.

navahgar commented on torch/csrc/jit/tensorexpr/ir_simplifier.cpp:
```diff
@@ -2404,6 +2404,92 @@ Stmt* TermExpander::mutate(const Block* v) {
+
+    if (block->nstmts() == 1) {
+      if (auto* cond = dynamic_cast<Cond*>(block->front())) {
+        Stmt* reordered = handleForCondReordering(v, cond);
```
What is this supposed to do? Please add your explanation as a comment here.

navahgar reviewed (commented):
Mostly LGTM. 

One meta comment. For all the not so trivial cases, it would be good to add your explanation of why that is being done as a comment. These would be necessary to understand the code when anyone reads this later.

huiguoo committed:
Update on "[nnc] Add simplifierUnderContext for simplification that needs context info: currently added for-stmt index var bounds info as context"


Differential Revision: [D29373315](https://our.internmc.facebook.com/intern/diff/D29373315)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373315).

huiguoo committed:
Update on "[nnc] Add simplifierUnderContext for simplification that needs context info: currently added for-stmt index var bounds info as context"


Differential Revision: [D29373315](https://our.internmc.facebook.com/intern/diff/D29373315)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29373315).
